File: Trees/MaxDepth.js
// Definition for a binary tree node.
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
 }
 
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    if (!root) return 0;

    const leftDepth = maxDepth(root.left);
    const rightDepth = maxDepth(root.right);

    return (leftDepth >= rightDepth ? leftDepth : rightDepth) + 1;
};

let root = new TreeNode(
    3,
    new TreeNode(9),
    new TreeNode(
        20, 
        new TreeNode(15),
        new TreeNode(17)
    )
);
console.log(maxDepth(root)); // 3

root = new TreeNode();
console.log(maxDepth(root)); // 1
console.log(maxDepth(null)); // 0

root = new TreeNode(
    1,
    null,
    new TreeNode(2)
);
console.log(maxDepth(root)); // 2

File: Trees/max_depth.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

#https://leetcode.com/problems/maximum-depth-of-binary-tree/
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        def max_depth_helper(root, depth=0):
            if not root:
                return depth

            return max(max_depth_helper(root.left, depth + 1), max_depth_helper(root.right, depth + 1))

        return max_depth_helper(root)



sol = Solution()
print(sol.maxDepth(None) == 0)
print(sol.maxDepth(TreeNode(1)) == 1)

root = TreeNode(1, None, TreeNode(2))
print(sol.maxDepth(root) == 2)

root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
print(sol.maxDepth(root) == 3)

